### ISSUE TITLE
Output layer question

### ISSUE THREAD

**Issue by bkj** [amzn/amazon-dsstne#178]
The output layer in these networks is often a bottleneck, because you have to do a `(batch_size, hidden_dim)` by `(hidden_dim, num_classes)` dense matrix multiplication.  It doesn't seem like you'd get a speed up _just_ by avoiding storing/multiplying by zeros -- are you doing any kinds of tricks here to reduce the cost of that operation?  

Thanks
~ Ben

**Comment by scottlegrand** [comment]
We have some ideas here based on approximate kNN methods.  Stay tuned.

**Comment by bkj** [comment]
Interesting -- are you thinking just for inference or for both inference and training?

**Comment by scottlegrand** [comment]
It's a no-brainer for inference, it's a science project for training.

**Comment by bkj** [comment]
Relevant code and paper, if you haven't seen it:
  https://github.com/rdspring1/LSH_DeepLearning
I don't think they did it on GPUs or big models, but maybe interesting

**Comment by scottlegrand** [comment]
Yep, I came up with roughly the same idea last Summer and then I read that
paper that fleshed it out even further last Fall.  My sparse input kernels
end up ~20x faster than SGEMM so their seeing more or less the same speedup
with LSH makes intuitive sense: one ends up memory-limited.  Older GPUs
would be a bear for this, but newer GPUs support arbitrary cuda streams and
they have much larger L2 caches so it shouldn't be all that hard to write.

**Comment by bkj** [comment]
Any updates on this?  I'm trying to find an example of a library that uses approximate kNN methods to speed up the output layer.
